### PR TITLE
New field 'body' to HTTP GET/GET-by-POST metadata

### LINKED ISSUE #271: Nested dynamic parameters in the body of HTTP external metadata POST requests
### Feature Request/How to

It would be nice if when retrieving metadata, we could do an HTTP post with nested dynamic data.

e.g. If we want to post the following JSON

```
{
  "request": {
    "key1": {
      "key2": {
        "key3": "192.168.0.10"
      }
    }
  }
}
```
Our metadata config would look like this and it works great:

```
metadata:
  - name: echo-api-post-json
    http:
      endpoint: http://talker-api.authorino.svc.cluster.local:3000/metadata?encoding=json
      method: POST
      contentType: application/json
      bodyParameters:
        - name: original_path
          valueFrom:
            authJSON: context.request.http.headers.x-forwarded-for
        - name: request
           value: 
             key1:
               key2:
                 key3: 192.168.0.10
```
However, how would we go about making "192.168.0.1" piece dynamic? e.g.
```
metadata:
  - name: echo-api-post-json
    http:
      endpoint: http://talker-api.authorino.svc.cluster.local:3000/metadata?encoding=json
      method: POST
      contentType: application/json
      bodyParameters:
        - name: original_path
          valueFrom:
            authJSON: context.request.http.headers.x-forwarded-for
        - name: request
           value: 
             key1:
               key2:
                 key3: 
                 valueFrom:
                   authJSON: context.request.http.headers.x-forwarded-for
```

Is this currently possible/supported?

### PR DESCRIPTION
**Pull request by guicassolato**
It adds a new field `spec.metadata.body`.

The new field is an alternative to `spec.metadata.bodyParameters` for defining raw HTTP bodies and ensure proper content encoding of JSON with multiple levels.

E.g., the following config:

```yaml
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: talker-api-protection
spec:
  hosts: ["talker-api-authorino.127.0.0.1.nip.io"]
  metadata:
  - name: echo-api-post-json
    http:
      endpoint: http://talker-api.default.svc.cluster.local:3000/metadata?encoding=json
      method: POST
      contentType: application/json
      body:
        valueFrom:
          authJSON: |
            \{"original_path":"{context.request.http.headers.x-forwarded-for}","request":\{"key1":\{"key2":\{"key3":"{context.request.http.headers.x-forwarded-for}"\}\}\}\}
```

will send a request:

```
POST /metadata?encoding=json HTTP/1.1
Content-Length: 81
Content-Type: application/json
Host: talker-api.default.svc.cluster.local:3000
User-Agent: Go-http-client/1.1
Accept-Encoding: gzip
Version: HTTP/1.1
(…request headers omitted)

{"original_path":"10.244.0.8","request":{"key1":{"key2":{"key3":"10.244.0.8"}}}}
```

A similar config to the above using `bodyParameters` instead would be:

```yaml
kubectl apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: talker-api-protection
spec:
  hosts: ["talker-api-authorino.127.0.0.1.nip.io"]
  metadata:
  - name: echo-api-post-json
    http:
      endpoint: http://talker-api.default.svc.cluster.local:3000/metadata?encoding=json
      method: POST
      contentType: application/json
      bodyParameters:
      - name: original_path
        valueFrom: { authJSON: context.request.http.headers.x-forwarded-for }
      - name: request
        valueFrom:
          authJSON: |
            \{"key1":\{"key2":\{"key3":"{context.request.http.headers.x-forwarded-for}"\}\}\}
```

In the latter however, 'request' is encoded as a JSON string, thus resulting in the following request sent to the metadata service:

```
POST /metadata?encoding=json HTTP/1.1
Content-Length: 92
Content-Type: application/json
Host: talker-api.default.svc.cluster.local:3000
User-Agent: Go-http-client/1.1
Accept-Encoding: gzip
Version: HTTP/1.1
(…request headers omitted)

{"original_path":"10.244.0.8","request":"{\"key1\":{\"key2\":{\"key3\":\"10.244.0.8\"}}}\n"}
```

Closes #271

### CODE REVIEW

File: install/crd/authorino.kuadrant.io_authconfigs.yaml [hunk 1]
```diff
@@ -988,10 +988,37 @@ spec:
...
+                                    @case:upper|lower, and @base64:encode|decode.'
+                                  type: string
+                              type: object
+                          type: object
                         bodyParameters:
                           description: Custom parameters to encode in the body of
-                            the HTTP request. Use it with method=POST; for GET requests,
-                            specify parameters using placeholders in the endpoint.
+                            the HTTP request. Superseded by 'body'; use either one
+                            or the other Use it with method=POST; for GET requests,
```
Comment (thomasmaas) [inline install/crd/authorino.kuadrant.io_authconfigs.yaml:1019]:
```suggestion
                            or the other. Use it with method=POST; for GET requests,
```

File: install/manifests.yaml [hunk 2]
```diff
@@ -1106,10 +1106,37 @@ spec:
...
+                                    @case:upper|lower, and @base64:encode|decode.'
+                                  type: string
+                              type: object
+                          type: object
                         bodyParameters:
                           description: Custom parameters to encode in the body of
-                            the HTTP request. Use it with method=POST; for GET requests,
-                            specify parameters using placeholders in the endpoint.
+                            the HTTP request. Superseded by 'body'; use either one
+                            or the other Use it with method=POST; for GET requests,
```
Comment (thomasmaas) [inline install/manifests.yaml:1137]:
```suggestion
                            or the other. Use it with method=POST; for GET requests,
```

File: install/crd/authorino.kuadrant.io_authconfigs.yaml (hunk 1 above)
Comment (guicassolato) [reply install/crd/authorino.kuadrant.io_authconfigs.yaml:1019]: Good catch. Thanks!

File: install/manifests.yaml (hunk 2 above)
Comment (guicassolato) [reply install/manifests.yaml]: Fixed.

**Review by eguzki — APPROVED** [review]

**Pull request merged**